sparse/_compressed/compressed.py

```diff
@@ -788,3 +799,87 @@ def _prune(self):
             self.indptr = indptr
         else:
             self.indices = self.indices[mask]
+
+
+class Compressed2d(GCXS):
+    def __init__(self, arg, shape=None, prune=False, fill_value=0):
+        if isinstance(arg, np.ndarray):
+            (arg, shape, _, fill_value) = _from_coo(COO(arg), self.compressed_axes)
+
+        elif isinstance(arg, COO):
+            (arg, shape, _, fill_value) = _from_coo(arg, self.compressed_axes)
+
+        elif isinstance(arg, type(self)):
+            (arg, shape, fill_value) = (
+                (arg.data, arg.indices, arg.indptr),
+                arg.shape,
+                arg.fill_value,
+            )
+
+        elif isinstance(arg, GCXS):
+            if len(arg.shape) != 2:
+                raise ValueError()
+            if arg.compressed_axes != self.compressed_axes:
+                arg = arg.change_compressed_axes(self.compressed_axes)
+            (arg, shape, fill_value) = (
+                (arg.data, arg.indices, arg.indptr),
+                arg.shape,
+                arg.fill_value,
+            )
+
+        if shape is None:
+            raise ValueError("missing `shape` argument")
+        elif len(shape) != 2:
+            raise ValueError(
+                f"{self.__name__} must be two dimensional, given shape {shape}."
+            )
+
+        check_compressed_axes(len(shape), self.compressed_axes)
+
+        self.data, self.indices, self.indptr = arg
+
+        if self.data.ndim != 1:
+            raise ValueError("data must be a scalar or 1-dimensional.")
+
+        self.shape = shape
+        self.fill_value = fill_value
+
+        if prune:
+            self._prune()
```

Comment:
Can we somehow re-use `super().__init__` here?

Reply:
A couple things stopped this for me:

* `self.compressed_axes = compressed_axes` in `GCXS.__init__`
* Being able to error once I knew the shape was wrong, instead of waiting until it's constructed

Reply:
Can we put the shape-checking logic above `super().__init__`?

Reply:
As concerns `compressed_axes`, that's constant anyway for these classes.

Reply:
Concretely, we can use it as:

```python
# Shape-checking goes here
super().__init__(..., compressed_axes=(0,)) # Or (1,) for CSC.
```

Reply:
`compressed_axes` is mutable for `GCXS`, but it really shouldn't be for `CSR` and `CSC`. Making a setter for `compressed_axes` that checks the value isn't different feels wrong to me. `compressed_axes` cannot be made immutable for `GCXS` because it's mutated by methods like `resize`.

I think I can do this with a little bit of introspection, but I'm not sure it's a good idea overall. For example, I thinks it's fairly simple to have a faster method for creating 2d compressed sparse arrays from 2d dense arrays, but am not sure this is the case for nd arrays.

There's specialization that can happen for the 2d case in the constructor, but sharing that code will add indirection for the nd case. This is a bit of a broader implementation question. For code that works with nd arrays, do they get a check at the front that says: `if 2d: fast_path_2d(...)` or does the 2d case just call the fast path directly?

Reply:
I've made the change in 01caa5c9781134fc4ec3e, but let me know what you think

Reply:
> I've made the change in [01caa5c](https://github.com/pydata/sparse/commit/01caa5c9781134fc4ec3e745162372290e2dfd94), but let me know what you think

We should make `compressed_axes` immutable, that's the solution I see. All these attributes should be immutable.

Reply:
What about `array.resize`? That's part of the "array interface" and can modify the shape and compressed axes inplace.

I also expect @daletovar may have some insight here.

Right now, `compressed_axes` is immutable for `CSR` and `CSC`, so any changes to the mutability of `GCXS` attributes is orthogonal. Could this be handled separately from this PR?

Reply:
> Right now, `compressed_axes` is immutable for `CSR` and `CSC`, so any changes to the mutability of `GCXS` attributes is orthogonal. Could this be handled separately from this PR?

Yes, I'm okay splitting it out. But we should create a tracking issue for all this. I'll do that.

> What about `array.resize`? That's part of the "array interface" and can modify the shape and compressed axes inplace.

I personally feel it's okay if we don't implement `resize` at all, I'm fine with `reshape` only.